### PR TITLE
Add maxHotplugRatio,maxCpuSockets & maxGuests VMI propagation tests

### PR DESCRIPTION
**Pull request by geetikakay**
Verify HCO maxHotplugRatio sets VMI maxSockets and maxGuest when instancetype doesn't define them explicitly.
Test VMI maxSockets/maxGuest values when HCO sets maxHotplugRatio, maxCpuSockets, and maxGuest in various combinations.

Assisted-by: Cursor

##### jira-ticket:
Based on the findings https://issues.redhat.com/browse/CNV-71380 and need of test coverage for edge cases and integration testing for all maxhotplugratio,maxcpusockets and maxguests with instancetypes.

## Summary by CodeRabbit

* **Tests**
  * Added a parameterized live-update test suite validating VM max sockets and max guest memory across multiple update scenarios.

* **Chores**
  * Added reusable fixtures for instancetype, test VM lifecycle, scenario handling and VM restart to support hotplug/live-update tests.
  * Introduced edge-case limit constants for cluster hotplug/cpu/memory.
  * Improved test utilities with stronger typing and a new helper to inspect controller revisions.

### CODE REVIEW

**Comment by geetikakay** [comment]
/wip
missing polarion id since polarion is not working

File: tests/infrastructure/instance_types/utils.py [hunk 1]
```diff
@@ -50,15 +67,49 @@ def assert_instance_revision_and_memory_update(
...
 def assert_secure_boot_mokutil_status(vm: VirtualMachineForTests) -> None:
+    """Assert Secure Boot is enabled via mokutil."""
     output = run_ssh_commands(host=vm.ssh_exec, commands=shlex.split("mokutil --sb-state"))[0].lower()
     assert "enabled" in output, f"Secure Boot is not enabled. Found: {output}"
 
 
 def assert_kernel_lockdown_mode(vm: VirtualMachineForTests) -> None:
+    """Assert kernel lockdown mode is not '[none]'."""
     output = run_ssh_commands(host=vm.ssh_exec, commands=shlex.split("cat /sys/kernel/security/lockdown"))[0]
     assert "[none]" not in output, f"Kernel lockdown mode is not '[none]'. Found: {output}"
```
[the remark this answers is not in the document]
Comment (geetikakay) [reply tests/infrastructure/instance_types/utils.py]: @coderabbitai Ack. Leaving as-is for now. The assert condition is clear enough and we prefer to keep the existing wording

File: tests/infrastructure/instance_types/conftest.py [hunk 2]
```diff
@@ -28,3 +38,67 @@ def base_vm_cluster_instancetypes(unprivileged_client):
...
+            data_source=rhel10_data_source_scope_session,
+        ),
+        os_flavor=OS_FLAVOR_RHEL,
+    ) as vm:
+        running_vm(vm=vm)
+        yield vm
+
+
+@pytest.fixture(scope="class")
+def vm_with_hco_live_update(request, hyperconverged_resource_scope_class, hotplug_test_vm):
```
Comment (vsibirsk) [inline tests/infrastructure/instance_types/conftest.py]: fixture should be 1 action only.
split into update hco fixture and restart vm

File: tests/infrastructure/instance_types/conftest.py [hunk 3]
```diff
@@ -28,3 +38,67 @@ def base_vm_cluster_instancetypes(unprivileged_client):
             label_selector=COMMON_INSTANCETYPE_SELECTOR,
         )
     )
+
+
+@pytest.fixture(scope="module")
+def u1_small_instancetype(unprivileged_client):
+    """Get the u1.small cluster instancetype."""
```
[the remark this answers is not in the document]
Comment (geetikakay) [reply tests/infrastructure/instance_types/conftest.py]: we never enforced but  fixtures under tests/infrastructure/conftest.py and tests/conftest.py  using it , so i kept them.

File: tests/infrastructure/instance_types/test_cluster_max_hotplug_ratio.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    CLUSTER_MAX_HOTPLUG_RATIO_MIN,
+)
+
+pytestmark = [
+    pytest.mark.tier3,
+    pytest.mark.rwx_default_storage,
+]
+
+SCENARIOS = [
+    pytest.param(
```
Comment (dshchedr) [inline tests/infrastructure/instance_types/test_cluster_max_hotplug_ratio.py]: don't create constant, add these params right in `@pytest.mark.parametrize`

File: tests/infrastructure/instance_types/conftest.py [hunk 5]
```diff
@@ -28,3 +38,62 @@ def base_vm_cluster_instancetypes(unprivileged_client):
...
+def hco_live_update_scenario(request, hyperconverged_resource_scope_class):
+    scenario = request.param
+    with ResourceEditorValidateHCOReconcile(
+        patches={
+            hyperconverged_resource_scope_class: {"spec": {"liveUpdateConfiguration": scenario["live_update_config"]}}
+        },
+        list_resource_reconcile=[KubeVirt],
+        wait_for_reconcile_post_update=True,
+    ):
+        yield scenario
```
[the remark this answers is not in the document]
Comment (geetikakay) [reply tests/infrastructure/instance_types/conftest.py]: added  yield here so the ResourceEditorValidateHCOReconcile context stays open while the test runs. If we return inside the the context exits immediately and restores HCO before the test body executes.

File: tests/infrastructure/instance_types/conftest.py [hunk 6]
```diff
@@ -28,3 +38,62 @@ def base_vm_cluster_instancetypes(unprivileged_client):
...
+        client=unprivileged_client,
+        name=U1_SMALL,
+    )
+
+
+@pytest.fixture()
+def s390x_hotplug_fail(is_s390x_cluster):
+    if is_s390x_cluster:
+        LOGGER.warning("Skipping hotplug tests on s390x: memory hotplug unsupported")
+        pytest.fail("s390x doesn't support memory hotplug")
```
[the remark this answers is not in the document]
Comment (geetikakay) [reply tests/infrastructure/instance_types/conftest.py]: removed fixture

File: tests/infrastructure/instance_types/conftest.py [hunk 7]
```diff
@@ -28,3 +34,57 @@ def base_vm_cluster_instancetypes(unprivileged_client):
             label_selector=COMMON_INSTANCETYPE_SELECTOR,
         )
     )
+
+
+@pytest.fixture(scope="module")
+def u1_small_instancetype(unprivileged_client):
```
[the remark this answers is not in the document]
Comment (geetikakay) [reply tests/infrastructure/instance_types/conftest.py]: https://issues.redhat.com/browse/CNV-79268

File: tests/infrastructure/instance_types/conftest.py [hunk 8]
```diff
@@ -28,3 +34,57 @@ def base_vm_cluster_instancetypes(unprivileged_client):
...
+    unprivileged_client,
+    u1_small_instancetype,
+    rhel10_data_source_scope_session,
+):
+    with VirtualMachineForTests(
+        name="test-hotplug-config-vm",
+        namespace=namespace.name,
+        client=unprivileged_client,
+        vm_instance_type=u1_small_instancetype,
+        vm_preference=VirtualMachineClusterPreference(
```
[the remark this answers is not in the document]
Comment (geetikakay) [reply tests/infrastructure/instance_types/conftest.py]: https://issues.redhat.com/browse/CNV-79268

File: tests/infrastructure/instance_types/test_cluster_max_hotplug_ratio.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+)
+
+pytestmark = [
+    pytest.mark.tier3,
+    pytest.mark.rwx_default_storage,
+]
+
+
+@pytest.mark.parametrize(
+    "hco_scenario",
```
[the remark this answers is not in the document]
Comment (RoniKishner) [reply tests/infrastructure/instance_types/test_cluster_max_hotplug_ratio.py]: Sounds like a good idea, it will make it easier to read

[the remark this answers is not in the document]
Comment (vsibirsk) [reply tests/infrastructure/instance_types/test_cluster_max_hotplug_ratio.py]: this creates redundant fixture. i'd prefer @RoniKishner 's initial solution

**Review by vsibirsk — CHANGES_REQUESTED** [review]

**Review by rnetser — CHANGES_REQUESTED** [review]

File: tests/infrastructure/instance_types/test_cluster_max_hotplug_ratio.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+from tests.infrastructure.instance_types.constants import (
+    CLUSTER_MAX_CPU_SOCKETS,
+    CLUSTER_MAX_GUEST,
+    CLUSTER_MAX_HOTPLUG_RATIO,
+    CLUSTER_MAX_HOTPLUG_RATIO_LARGE,
+    CLUSTER_MAX_HOTPLUG_RATIO_MIN,
+)
+
+pytestmark = pytest.mark.rwx_default_storage
```
[the remark this answers is not in the document]
Comment (geetikakay) [reply tests/infrastructure/instance_types/test_cluster_max_hotplug_ratio.py]: yeah picked name based on parameters and it's maxHotplugratio. will rename

File: tests/infrastructure/instance_types/test_cluster_max_hotplug_ratio.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
+# flake8: noqa: PID001 - Class-level parametrize; Polarion IDs are in pytest.param marks
```
[the remark this answers is not in the document]
Comment (rnetser) [reply tests/infrastructure/instance_types/test_cluster_max_hotplug_ratio.py]: so this line can be removed. right?

File: tests/infrastructure/instance_types/conftest.py [hunk 12]
```diff
@@ -28,3 +34,52 @@ def base_vm_cluster_instancetypes(unprivileged_client):
...
+@pytest.fixture(scope="module")
+def u1_small_instancetype(unprivileged_client):
+    return VirtualMachineClusterInstancetype(
+        client=unprivileged_client,
+        name=U1_SMALL,
+    )
+
+
+@pytest.fixture(scope="class")
+def hotplug_test_vm(
```
[the remark this answers is not in the document]
Comment (geetikakay) [reply tests/infrastructure/instance_types/conftest.py]: yeah picked name based on parameters and it's maxHotplugratio. will rename
